pkg/controller/client_builder_dynamic.go

```diff
@@ -0,0 +1,214 @@
+			authClient:         t.AuthenticationClient,
+		}, dur)
+		configCopy.WrapTransport = transport.TokenSourceWrapTransport(cachedTokenSource)
+
+		t.set(saName, configCopy.WrapTransport)
+	}
+
+	return &configCopy, nil
+}
+
+func (t *DynamicControllerClientBuilder) ConfigOrDie(name string) *restclient.Config {
+	clientConfig, err := t.Config(name)
+	if err != nil {
+		klog.Fatal(err)
+	}
+	return clientConfig
+}
+
+func (t *DynamicControllerClientBuilder) Client(name string) (clientset.Interface, error) {
+	clientConfig, err := t.Config(name)
+	if err != nil {
+		return nil, err
+	}
+	return clientset.NewForConfig(clientConfig)
+}
+
+func (t *DynamicControllerClientBuilder) ClientOrDie(name string) clientset.Interface {
+	client, err := t.Client(name)
+	if err != nil {
+		klog.Fatal(err)
+	}
+	return client
+}
+
+func (t *DynamicControllerClientBuilder) getOrCreateServiceAccount(name string) (*v1.ServiceAccount, error) {
+	sa, err := t.CoreClient.ServiceAccounts(t.Namespace).Get(name, metav1.GetOptions{})
+	if err == nil {
+		return sa, nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	// Create the namespace if we can't verify it exists.
+	// Tolerate errors, since we don't know whether this component has namespace creation permissions.
+	if _, err := t.CoreClient.Namespaces().Get(t.Namespace, metav1.GetOptions{}); apierrors.IsNotFound(err) {
+		if _, err = t.CoreClient.Namespaces().Create(&v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: t.Namespace}}); err != nil {
+			klog.Warningf("create non-exist namespace %s failed:%v", t.Namespace, err)
+		}
+	}
+
+	// Create the service account
+	sa, err = t.CoreClient.ServiceAccounts(t.Namespace).Create(&v1.ServiceAccount{ObjectMeta: metav1.ObjectMeta{Namespace: t.Namespace, Name: name}})
+	if apierrors.IsAlreadyExists(err) {
+		// If we're racing to init and someone else already created it, re-fetch
+		return t.CoreClient.ServiceAccounts(t.Namespace).Get(name, metav1.GetOptions{})
+	}
+	return sa, err
+}
+
+func (t *DynamicControllerClientBuilder) get(key string) (func(http.RoundTripper) http.RoundTripper, bool) {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+	rt, ok := t.roundTripperFuncMap[key]
+	return rt, ok
+}
+
+func (t *DynamicControllerClientBuilder) set(key string, rt func(http.RoundTripper) http.RoundTripper) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	t.roundTripperFuncMap[key] = rt
+}
+
+type tokenSourceImpl struct {
+	clientConfig       *restclient.Config
+	namespace          string
+	serviceAccountName string
+	coreClient         v1core.CoreV1Interface
+	authClient         v1authentication.AuthenticationV1Interface
+}
+
+func (ts *tokenSourceImpl) Token() (*oauth2.Token, error) {
+	var retTokenRequest *v1authenticationapi.TokenRequest
+
+	if err := wait.Poll(100*time.Millisecond, 5*time.Second, func() (bool, error) {
+		tr, inErr := ts.coreClient.ServiceAccounts(ts.namespace).CreateToken(ts.serviceAccountName, &v1authenticationapi.TokenRequest{
```

Comment:
I don't think we should do this:

- This `Token() `function should only take care about token generate. It should consider service account missing as an unexpected situation.

- This function has no context to know when it is required to recreate a serviceaccount

- In legacy client builder, it would not try recover the serviceaccount neither.

Reply:
The client builder is creating the service account at init time, and the Token() function will permanently fail if the service account is missing at token renew time. What are the pros and cons of having it rebuild the service account object?

> This function has no context to know when it is required to recreate a serviceaccount

A NotFound error followed by a check of the service account also returning NotFound is a pretty good hint :)

> In legacy client builder, it would not try recover the serviceaccount neither.

The clients produced by the legacy client builder were not resilient to key rotations or token revocations. I think we can do better here.